File: vanilla-nodejs-backend/src/router/route.ts
import { IncomingMessage, ServerResponse } from "http"
import { signIn, signUp } from "../controllers/auth-controller"
import { HttpStatusCode } from "axios";
import { isUrlMatching } from "../utils/utils";
import url from 'url'
import { storesController, uniqueStoreController } from "../controllers/store-controller";
import { billboardsController, uniqueBillboardController } from "../controllers/billboard-controller";
import { categoriesController, uniqueCategoryController } from "../controllers/category-controller";
import { sizesController, uniqueSizeController } from "../controllers/size-controller";
import { colorsController, uniqueColorController } from "../controllers/color-controller";
import { productsController, uniqueProductController } from "../controllers/product-controller";
import { ordersController } from "../controllers/order-controller";

const routes = [
    {path : '/auth/signup' , handler : signUp},
    {path : '/auth/signin' , handler : signIn},
    {path : '/store/' , handler : storesController},
    {path : '/store/:storeId' , handler : uniqueStoreController},
    {path : '/store/:storeId/billboards' , handler : billboardsController},
    {path : '/store/:storeId/billboards/:billboardId' , handler : uniqueBillboardController},
    {path : '/store/:storeId/categories/' , handler : categoriesController},
    {path : '/store/:storeId/categories/:categoryId' , handler : uniqueCategoryController},
    {path : '/store/:storeId/sizes' , handler : sizesController},
    {path : '/store/:storeId/sizes/:sizeId' , handler : uniqueSizeController},
    {path : '/store/:storeId/colors' , handler : colorsController},
    {path : '/store/:storeId/colors/:colorId' , handler : uniqueColorController},
    {path : '/store/:storeId/products' , handler : productsController},
    {path : '/store/:storeId/products/:productId' , handler : uniqueProductController},
    {path : '/store/:storeId/orders' , handler : ordersController},
]

export const router = async (req:IncomingMessage,res:ServerResponse)=>{
    const parsedUrl = url.parse(req.url as string,true)
    const pathname = parsedUrl.pathname
    const route = routes.find( r => isUrlMatching(pathname as string,r.path))
    if(route) {
        try {
            await route.handler(req,res)
        } catch (error) {
            res.writeHead(HttpStatusCode.InternalServerError)
            res.end("InternalServerError")
        }
    } else {
        res.writeHead(HttpStatusCode.NotFound)
        res.end("Not Found")
    }
}

File: golang-http-backend/routes/route.go
package routes

import (
	"connection-to-mongo/project/controller"
	"connection-to-mongo/project/middleware"

	"github.com/gorilla/mux"
)

func AddAuthorizationRoute(r *mux.Router) {
	auth := r.PathPrefix("/auth").Subrouter()
	auth.HandleFunc("/signup", controller.SignUp).Methods("POST")
	auth.HandleFunc("/signin", controller.SignIn).Methods("POST")

}

func AddStoreRoute(r *mux.Router) {
	store := r.PathPrefix("/store").Subrouter()
	store.Handle("/", middleware.AuthMiddleware(controller.CreateStoreController)).Methods("POST")
	store.HandleFunc("/", controller.GetStores).Methods("GET")
	store.HandleFunc("/{storeId}", controller.GetStore).Methods("GET")

}

func AddBillboardRoute(r *mux.Router) {

	billboard := r.PathPrefix("/store/{storeId}/billboards").Subrouter()
	billboard.Handle("", middleware.AuthMiddleware(controller.BillboardsController)).Methods("POST", "GET")
	billboard.Handle("/{billboardId}", middleware.AuthMiddleware(controller.UniqueBillboardController)).Methods("GET", "PATCH", "DELETE")

}

func AddCategoryRoute(r *mux.Router) {

	billboard := r.PathPrefix("/store/{storeId}/categories").Subrouter()
	billboard.Handle("", middleware.AuthMiddleware(controller.CategoriesController)).Methods("POST", "GET")
	billboard.Handle("/{categoryId}", middleware.AuthMiddleware(controller.UniqueCategoryController)).Methods("GET", "PATCH", "DELETE")

}

func AddSizeRoute(r *mux.Router) {

	billboard := r.PathPrefix("/store/{storeId}/sizes").Subrouter()
	billboard.Handle("", middleware.AuthMiddleware(controller.SizesController)).Methods("POST", "GET")
	billboard.Handle("/{sizeId}", middleware.AuthMiddleware(controller.UniqueSizeController)).Methods("GET", "PATCH", "DELETE")

}

func AddColorRoute(r *mux.Router) {

	billboard := r.PathPrefix("/store/{storeId}/colors").Subrouter()
	billboard.Handle("", middleware.AuthMiddleware(controller.ColorsController)).Methods("POST", "GET")
	billboard.Handle("/{colorId}", middleware.AuthMiddleware(controller.UniqueColorController)).Methods("GET", "PATCH", "DELETE")

}

func AddProductRoute(r *mux.Router) {

	billboard := r.PathPrefix("/store/{storeId}/products").Subrouter()
	billboard.Handle("", middleware.AuthMiddleware(controller.ProductsController)).Methods("POST", "GET")
	billboard.Handle("/{productId}", middleware.AuthMiddleware(controller.UniqueProductController)).Methods("GET", "PATCH", "DELETE")

}

func AddOrderRoute(r *mux.Router) {

	billboard := r.PathPrefix("/store/{storeId}/orders").Subrouter()
	billboard.Handle("", middleware.AuthMiddleware(controller.OrdersController)).Methods("POST", "GET")
	// billboard.Handle("/{orderId}", middleware.AuthMiddleware(controller.UniqueProductController)).Methods("GET", "PATCH", "DELETE")
}
